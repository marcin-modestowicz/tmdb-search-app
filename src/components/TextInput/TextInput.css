.root {
  display: flex;
  align-items: center;
  border: 1px solid dimGray;
  border-radius: 0.25em;
}

.root.active {
  border-color: royalBlue;
}

.root svg {
  margin-right: 0.5em;
}

.input {
  display: inline-block;
  padding: 0 0.5em;
  font-family: sans-serif;
  color: dimGray;
  font-size: 1em;
  line-height: 2em;
  box-sizing: border-box;
  flex: 1;
  border: none;
  outline: none;
  background: transparent;
  -webkit-appearance: none;
  box-sizing: border-box;
}
