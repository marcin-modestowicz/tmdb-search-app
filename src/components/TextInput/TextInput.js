// @flow
import React, { Component } from "react";
import { observable, action } from "mobx";
import { observer } from "mobx-react";
import cn from "classnames";
import styles from "./TextInput.css";

/* global SyntheticInputEvent, React$Element */

type Props = {
  value: ?string,
  placeholder?: string,
  onChange?: (value: string) => void,
  onBlur?: () => void,
  onFocus?: () => void,
  onMouseEnter?: (name?: string) => void,
  onMouseLeave?: (name?: string) => void,
  autocomplete?: string,
  autocorrect?: string,
  autocapitalize?: string,
  spellcheck?: boolean,
  autoFocus?: boolean,
  icon?: ?React$Element<any>
};

@observer
class TextInput extends Component<Props> {
  props: Props;

  static defaultProps = {
    placeholder: "",
    autocomplete: "on",
    autocapitalize: "on",
    autocorrect: "on",
    spellcheck: true,
    autoFocus: false
  };

  @observable isFocused = false;

  handleChange = (event: SyntheticInputEvent<*>) => {
    const value = event.target.value;

    if (this.props.onChange && value !== this.props.value) {
      this.props.onChange(value);
    }
  };

  @action
  handleFocus = () => {
    this.isFocused = true;

    if (this.props.onFocus) {
      this.props.onFocus();
    }
  };

  @action
  handleBlur = () => {
    this.isFocused = false;

    if (this.props.onBlur) {
      this.props.onBlur();
    }
  };

  render() {
    const {
      value,
      placeholder,
      onMouseEnter,
      onMouseLeave,
      autocapitalize,
      autocorrect,
      autocomplete,
      spellcheck,
      autoFocus,
      icon
    } = this.props;

    return (
      <div className={cn(styles.root, { [styles.active]: this.isFocused })}>
        <input
          type="text"
          value={value}
          placeholder={placeholder}
          className={styles.input}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          onFocus={this.handleFocus}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          autoCapitalize={autocapitalize}
          autoCorrect={autocorrect}
          autoComplete={autocomplete}
          spellCheck={spellcheck}
          autoFocus={autoFocus}
          tabIndex="0"
        />
        <div className={styles.icon}>{icon}</div>
      </div>
    );
  }
}

export default TextInput;
