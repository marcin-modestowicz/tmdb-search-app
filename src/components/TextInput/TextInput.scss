.root {
  display: flex;
  align-items: center;
  justify-content: space-between;
  border: 1px solid dimGray;
  border-radius: 0.25em;
  font-family: sans-serif;

  &.active {
    border-color: royalBlue;
  }

  .input {
    display: inline-block;
    margin: 0;
    padding: 0 0.5em;
    color: dimGray;
    font-size: 1em;
    line-height: 2em;
    box-sizing: border-box;
    min-width: 2em;
    text-overflow: ellipsis;
    border: none;
    outline: none;
    background: transparent;
    -webkit-appearance: none;
    box-sizing: border-box;
  }

  .icon {
    margin-right: 0.5em;
  }
}
