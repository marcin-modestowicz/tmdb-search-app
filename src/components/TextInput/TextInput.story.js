import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import { withKnobs, text, boolean } from "@storybook/addon-knobs";
import TextInput from "./TextInput";

storiesOf("TextInput", module)
  .addDecorator(withKnobs)
  .add("simple", () => (
    <TextInput
      value={text("Value", "Sample text")}
      onChange={action("value changed")}
      onBlur={action("blur")}
      onFocus={action("focus")}
      onMouseEnter={action("mouse enter")}
      onMouseLeave={action("mouse leave")}
      autocomplete={boolean("Autocomplete", true)}
      autocorrect={boolean("Autocorrect", true)}
      autocapitalize={boolean("Autocapitalize", true)}
      spellcheck={boolean("Spellcheck", true)}
      autoFocus={boolean("AutoFocus", true)}
    />
  ))
  .add("with icon", () => (
    <TextInput
      value={text("Value", "Sample text")}
      onChange={action("value changed")}
      onBlur={action("blur")}
      onFocus={action("focus")}
      onMouseEnter={action("mouse enter")}
      onMouseLeave={action("mouse leave")}
      autocomplete={boolean("Autocomplete", true)}
      autocorrect={boolean("Autocorrect", true)}
      autocapitalize={boolean("Autocapitalize", true)}
      spellcheck={boolean("Spellcheck", true)}
      autoFocus={boolean("AutoFocus", true)}
      icon={
        <svg width="16" height="16" viewBox="0 0 300 300">
          <path
            d="M 150,0 a 150,150 0 0,1 106.066,256.066 l -35.355,-35.355 a -100,-100 0 0,0 -70.711,-170.711 z"
            fill="#76f19a"
          >
            <animateTransform
              attributeName="transform"
              attributeType="XML"
              type="rotate"
              from="0 150 150"
              to="360 150 150"
              begin="0s"
              dur=".5s"
              fill="freeze"
              repeatCount="indefinite"
            />
          </path>
        </svg>
      }
    />
  ));
