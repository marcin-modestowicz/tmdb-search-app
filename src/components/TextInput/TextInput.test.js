// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import TextInput from "./TextInput";

describe("TextInput component", () => {
  let textInput;
  let onChange;
  let onFocus;
  let onBlur;
  let onMouseEnter;
  let onMouseLeave;

  beforeEach(() => {
    onChange = jest.fn();
    onFocus = jest.fn();
    onBlur = jest.fn();
    onMouseEnter = jest.fn();
    onMouseLeave = jest.fn();
    textInput = shallow(
      <TextInput
        value="one"
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      />
    );
  });

  test("should render inactive", () => {
    expect(toJson(textInput)).toMatchSnapshot();
  });

  test("should render active", () => {
    textInput.find("input").simulate("focus");

    expect(toJson(textInput)).toMatchSnapshot();
  });

  test("should render with icon", () => {
    textInput = shallow(
      <TextInput
        value="one"
        onChange={onChange}
        icon={
          <svg>
            <rect width="30" height="10" />
          </svg>
        }
      />
    );

    expect(toJson(textInput)).toMatchSnapshot();
  });

  test("onChange handler should be triggered with text value", () => {
    textInput.find("input").simulate("change", { target: { value: "text" } });

    expect(onChange).toHaveBeenCalledWith("text");
  });

  test("onFocus handler should be triggered", () => {
    textInput.find("input").simulate("focus");
    expect(onFocus).toHaveBeenCalled();
  });

  test("onBlur handler should be triggered", () => {
    textInput.find("input").simulate("blur");
    expect(onBlur).toHaveBeenCalled();
  });

  test("onMouseEnter handler should be triggered", () => {
    textInput.find("input").simulate("mouseEnter");
    expect(onMouseEnter).toHaveBeenCalled();
  });

  test("onMouseLeave handler should be triggered", () => {
    textInput.find("input").simulate("mouseLeave");
    expect(onMouseLeave).toHaveBeenCalled();
  });
});
