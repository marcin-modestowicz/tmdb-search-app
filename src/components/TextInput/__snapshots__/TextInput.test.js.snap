// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput component should render active 1`] = `
<div
  className="root active"
>
  <input
    autoCapitalize="on"
    autoComplete="on"
    autoCorrect="on"
    autoFocus={false}
    className="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    placeholder=""
    spellCheck={true}
    tabIndex="0"
    type="text"
    value="one"
  />
  <div
    className="icon"
  />
</div>
`;

exports[`TextInput component should render inactive 1`] = `
<div
  className="root"
>
  <input
    autoCapitalize="on"
    autoComplete="on"
    autoCorrect="on"
    autoFocus={false}
    className="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    placeholder=""
    spellCheck={true}
    tabIndex="0"
    type="text"
    value="one"
  />
  <div
    className="icon"
  />
</div>
`;

exports[`TextInput component should render with icon 1`] = `
<div
  className="root"
>
  <input
    autoCapitalize="on"
    autoComplete="on"
    autoCorrect="on"
    autoFocus={false}
    className="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    spellCheck={true}
    tabIndex="0"
    type="text"
    value="one"
  />
  <div
    className="icon"
  >
    <svg>
      <rect
        height="10"
        width="30"
      />
    </svg>
  </div>
</div>
`;
